### ISSUE TITLE
Release process changes

### ISSUE THREAD

**Issue by vrurg** [perl6/problem-solving#54]
This one is simple and purely rakudo-oriented.

The current release process is basically based on pinpointing a certain commit in master and calling it `the release`. The process is basically involves a kind of development freeze when major changes are better be postponed because they may cause breakages and block the release.

I would like to propose cutting releases by creating release branches based on the most stable commits. A release branch would only incorporate changes fixing release blockers.

Advantages of this method also include less pressure put on a release manager by giving him more flexibility on when and how to perform all the release tasks. Also, I'd expect it to be fewer release skips because of this.

**Comment by AlexDaniel** [comment]
It sounds great in theory but I'm not sure if it's going to work in practice. Problem is, blockers often require some effort to get them fixed, which means having more eyes on them is a good thing. So in some way the “development freeze” gets more people to work on the blockers.

> Advantages of this method also include less pressure put on a release manager by giving him more flexibility on when and how to perform all the release tasks

Yes, which is why most rakudo releases are cut from a release branch. So in some way what you're saying is already in use. However, I only do that once all blockers are resolved (meaning only release-related commits are required, e.g. changelog, release announcement, etc.).

**Comment by AlexDaniel** [comment]
Another thing worth mentioning is that while most of rakudo releases were done from a release branch, this is not a standard practice for MoarVM, meaning that at times we were kinda blocked by that. I'm all for consistifying the release processes across all projects, regardless of how much people want to keep things separate.

**Comment by vrurg** [comment]
> Yes, which is why most rakudo releases are cut from a release branch. So in some way what you're saying is already in use. However, I only do that once all blockers are resolved (meaning only release-related commits are required, e.g. changelog, release announcement, etc.).

I'm thinking of myself in this situation and it doesn't matter to me if there is a release branch or there isn't. What matters is an issue assigned and tagged as a blocker which I want to be closed. The work is done on `master` in either case. It would be up to the release manager to cherry pick the fix into the release.

**Comment by jnthn** [comment]
> Another thing worth mentioning is that while most of rakudo releases were done from a release branch, this is not a standard practice for MoarVM, meaning that at times we were kinda blocked by that.

Typically, we try to keep `master` always releasable and do new development that might be higher-risk in a branch. Thus, complex long-running features will be done in a branch (such as the various stages of the partial escape analysis). Even smaller but still somewhat risky things - especially implementing new optimizations - in the time leading up to a release, are encouraged to be put in a branch. Merges of such tend to be done just after a release, to maximize testing/fixing time.

**Comment by vrurg** [comment]
> Typically, we try to keep `master` always releasable and do new development that might be higher-risk in a branch. Thus, complex long-running features will be done in a branch (such as the various stages of the partial escape analysis). Even smaller but still somewhat risky things - especially implementing new optimizations - in the time leading up to a release, are encouraged to be put in a branch. Merges of such tend to be done just after a release, to maximize testing/fixing time.

That's why my proposal. I think that could improve the pace of development. Perhaps my thinking is distorted by the fact that I often end up in a situation where changing one thing requires another one to be straighten out. But I hesitate to merge changes because of an upcoming new release.

So, generally – yes, the proposal would reduce the overall level of `master`'s releaseability. But is it really that bad with somewhat faster development and more regular releases?

That's basically all my reasoning.

**Comment by AlexDaniel** [comment]
I'd wait for @kawaii to finish the proposal they are working on ([link](https://github.com/perl6/problem-solving/issues/53#issuecomment-507667742)). I think the picture will be a bit clearer once there's something.

**Comment by lizmat** [comment]
I think this issue has gone stale, since the release process got much more streamlined in the past 5 years.  Please create a new issue if appropriate.
